src/infql.jl

```diff
@@ -378,3 +378,216 @@ function LazyBandedMatrices._SymTridiagonal(::Tuple{TriangularLayout{'L', 'N', P
+        end
+        Ls = Ll
+        N = 2*N
+    end
+    F = ql(A[1:(N÷2),1:(N÷2)])
+    return (F.factors[1:50,1:50], F.τ[1:50])
+end
+
+# Resize and filling functions for cached implementation
+function resizedata!(K::AdaptiveQLFactors, nm::Integer)
+    νμ = K.datasize[1]
+    if nm > νμ
+        olddata = copy(K.data)
+        K.data = similar(K.data, nm, nm)
+        K.data[axes(olddata)...] = olddata
+        inds = νμ:nm
+        cache_filldata!(K, inds)
+        K.datasize = size(K.data)
+    end
+    K
+end
+
+function resizedata!(K::AdaptiveQLTau, nm::Integer)
+    νμ = K.datasize
+    if nm > νμ
+        resize!(K.data,νμ)
+        cache_filldata!(K, νμ:nm)
+        K.datasize = size(K.data,1)
+    end
+    K
+end
+
+function cache_filldata!(A::AdaptiveQLFactors{T}, inds::UnitRange{Int}) where T
+    j = maximum(inds)
+    maxN = 1000*j # Prevent runaway loop
+    Lerr = one(T)
+    N = j
+    checkinds = max(1,j-bandwidth(A.M,1)-bandwidth(A.M,2))
+    @inbounds Ls = ql(A.M[checkinds:N,checkinds:N]).L[2:j-checkinds+1,2:j-checkinds+1]
+    @inbounds while Lerr > A.tol
+        # compute QL for small finite section and large finite section
+        Ll = ql(A.M[checkinds:2N,checkinds:2N]).L[2:j-checkinds+1,2:j-checkinds+1]
+        # compare bottom right sections and stop if desired level of convergence achieved
+        Lerr = norm(Ll-Ls,2)
+        if N == maxN
+            error("Reached max. iterations in finite section QL without convergence to desired tolerance.")
+        end
+        Ls = Ll
+        N = 2*N
+    end
+    A.data = ql(A.M[1:(N÷2),1:(N÷2)]).factors[1:j,1:j]
+end
+
+
+function cache_filldata!(A::AdaptiveQLTau{T}, inds::UnitRange{Int}) where T
+    j = maximum(inds)
+    maxN = 1000*j
+    Lerr = one(T)
+    N = j
+    checkinds = max(1,j-bandwidth(A.M,1)-bandwidth(A.M,2))
+    @inbounds Ls = ql(A.M[checkinds:N,checkinds:N]).L[2:j-checkinds+1,2:j-checkinds+1]
+    @inbounds while Lerr > A.tol
+        # compute QL for small finite section and large finite section
+        Ll = ql(A.M[checkinds:2N,checkinds:2N]).L[2:j-checkinds+1,2:j-checkinds+1]
+        # compare bottom right sections and stop if desired level of convergence achieved
+        Lerr = norm(Ll-Ls,2)
+        if N == maxN
+            error("Reached max. iterations in finite section QL without convergence to desired tolerance.")
+        end
+        Ls = Ll
+        N = 2*N
+    end
+    A.data = ql(A.M[1:(N÷2),1:(N÷2)]).τ[1:j]
+end
+
+function getindex(K::AdaptiveQLFactors, k::Integer, j::Integer)
+    resizedata!(K, max(k,j))
+    K.data[k, j]
+end
+function getindex(K::AdaptiveQLFactors, k::Integer, jr::UnitRange{Int})
+    resizedata!(K, max(k,maximum(jr)))
+    K.data[k, jr]
+end
+function getindex(K::AdaptiveQLFactors, kr::UnitRange{Int}, j::Integer)
+    resizedata!(K, max(j,maximum(kr)))
+    K.data[kr, j]
+end
+function getindex(K::AdaptiveQLFactors, kr::UnitRange{Int}, jr::UnitRange{Int})
```

Comment:
Do we need this function? I'm pretty sure being `<: AbstractCachedArray` gives us this for free.

Reply:
So the problem here if I remember correctly is that the default functions in `::AbstractCachedArray` mean that `getindex` with `UnitRange` just calls each element individually despite it being a banded matrix, resulting in a dense block.

Without my getindex:
```julia
julia> F.factors[1:10,1:10]
10×10 Matrix{Float64}:
 -0.38233   0.70025    0.0        0.0        0.0        0.0        0.0        0.0        0.0        0.0
 -1.20329  -1.06416    0.3598     0.0        0.0        0.0        0.0        0.0        0.0        0.0
 -0.63712  -2.69889   -3.13912    0.198509   0.0        0.0        0.0        0.0        0.0        0.0
  0.0      -0.763932  -2.0       -2.61803    0.198509   0.0        0.0        0.0        0.0        0.0
  0.0       0.0       -0.381966  -2.0       -2.61803    0.198509   0.0        0.0        0.0        0.0
  0.0       0.0        0.0       -0.381966  -2.0       -2.61803    0.198509   0.0        0.0        0.0
  0.0       0.0        0.0        0.0       -0.381966  -2.0       -2.61803    0.198509   0.0        0.0
  0.0       0.0        0.0        0.0        0.0       -0.381966  -2.0       -2.61803    0.198509   0.0
  0.0       0.0        0.0        0.0        0.0        0.0       -0.381966  -2.0       -2.61803    0.198509
  0.0       0.0        0.0        0.0        0.0        0.0        0.0       -0.381966  -2.0       -2.61803
```

With my own getindex:
```julia
julia> F.factors[1:10,1:10]
10×10 BandedMatrix{Float64} with bandwidths (2, 1):
 -0.38233   0.70025     ⋅          ⋅          ⋅          ⋅          ⋅          ⋅          ⋅          ⋅ 
 -1.20329  -1.06416    0.3598      ⋅          ⋅          ⋅          ⋅          ⋅          ⋅          ⋅ 
 -0.63712  -2.69889   -3.13912    0.198509    ⋅          ⋅          ⋅          ⋅          ⋅          ⋅ 
   ⋅       -0.763932  -2.0       -2.61803    0.198509    ⋅          ⋅          ⋅          ⋅          ⋅ 
   ⋅         ⋅        -0.381966  -2.0       -2.61803    0.198509    ⋅          ⋅          ⋅          ⋅ 
   ⋅         ⋅          ⋅        -0.381966  -2.0       -2.61803    0.198509    ⋅          ⋅          ⋅ 
   ⋅         ⋅          ⋅          ⋅        -0.381966  -2.0       -2.61803    0.198509    ⋅          ⋅ 
   ⋅         ⋅          ⋅          ⋅          ⋅        -0.381966  -2.0       -2.61803    0.198509    ⋅ 
   ⋅         ⋅          ⋅          ⋅          ⋅          ⋅        -0.381966  -2.0       -2.61803    0.198509
   ⋅         ⋅          ⋅          ⋅          ⋅          ⋅          ⋅        -0.381966  -2.0       -2.61803

```
and it's faster too.

So not only do the standard methods break bandedness which is a no go, it also calls entries individually when supplying a unitrange (the most common way of calling a matrix block imo) which means that it is step by step filled in instead of filing in with the maximum once and then just calling the entries. If we want to not specify our own getindex here this needs to be changed further down, I think.

I believe way back when cached arrays were first implemented we had a chat about this step by step increase versus just increasing to the max value and the conclusion was that my case in SemiclassicalOPs was a special case  anyway and so I should just implement a new getindex, resizearray combo for it. But maybe now this is basic enough to consider it a bug.

Reply:
If `MemoryLayout(F) <: AbstractBandedLayout` then it will work correctly.

Reply:
Let me give this a shot

Reply:
but we already have `MemoryLayout(::AdaptiveQLFactors) = LazyBandedLayout()`, seems to me that should do it?

In practice I get

```julia
julia> MemoryLayout(F.factors)
LazyBandedMatrices.LazyBandedLayout()

julia> F.factors[1:10,1:10]
10×10 Matrix{Float64}:
 -0.38233   0.70025    0.0        0.0        0.0        0.0        0.0        0.0        0.0        0.0
 -1.20329  -1.06416    0.3598     0.0        0.0        0.0        0.0        0.0        0.0        0.0
 -0.63712  -2.69889   -3.13912    0.198509   0.0        0.0        0.0        0.0        0.0        0.0
  0.0      -0.763932  -2.0       -2.61803    0.198509   0.0        0.0        0.0        0.0        0.0
  0.0       0.0       -0.381966  -2.0       -2.61803    0.198509   0.0        0.0        0.0        0.0
  0.0       0.0        0.0       -0.381966  -2.0       -2.61803    0.198509   0.0        0.0        0.0
  0.0       0.0        0.0        0.0       -0.381966  -2.0       -2.61803    0.198509   0.0        0.0
  0.0       0.0        0.0        0.0        0.0       -0.381966  -2.0       -2.61803    0.198509   0.0
  0.0       0.0        0.0        0.0        0.0        0.0       -0.381966  -2.0       -2.61803    0.198509
  0.0       0.0        0.0        0.0        0.0        0.0        0.0       -0.381966  -2.0       -2.61803
```

Reply:
Ah, wait you said `MemoryLayout(F)`

Reply:
But actually we probably just want these (with `UnitRange`) added to LazyArrays.jl

Reply:
> In practice I get

that's definitely a bug....

Reply:
It's been a bug for a too long then :laughing: I can make a pull request into LazyArrays with these getindex variants if you want

Reply:
Started a PR: https://github.com/JuliaArrays/LazyArrays.jl/pull/248

Reply:
Ok I moved the getindex functionality out from here and into that PR for LazyArrays.